### PR TITLE
Add checkpoint, rollback to to bank

### LINKED ISSUE #1426: new feature: forking bank
To track trees of forks in the TPU and perform rollback, the Bank needs to be fork-able.  The bottom bank represents the last rollback point.

new functions needed?  new type needed?
BankFork::new(bank: Arc<Bank>, tick: Entry)
BankFork::process_entries() -> Result<()>
BankFork::collapse() -> (Bank, Vec<Entry>)

Internally Bank needs to understand how to chain account state.

Internally Bank needs a concept of zero balance.

### PR DESCRIPTION
**Pull request by rob-solana**
**Problem**
Bank does not implement checkpoint, rollback, and collapse to follow forks in the ledger.

**Summary of Changes**
Add linked lists of accounts.
checkpoint == push current accounts to "parent"
rollback == drop head, pull parent into head
collapse(depth) == recurse by parent, if node is depth deep or more, eat your parent

Fixes #1426

### CODE REVIEW

**Comment by garious** [comment]
I like the approach - to make each of bank's fields into a forkable one. Then `Bank` will ultimately implement the same `collapse()`, `fork()` and `rollback()` methods by calling the same method on each field?

**Comment by rob-solana** [comment]
I was going to have the whole bank fork, but ran into lots of issues.  This approach is somewhat micro, but matches how the bank is currently structured.

**Comment by garious** [comment]
Needs tests

**Comment by rob-solana** [comment]
more tests to come, I think the tree I'm barking up looks right-ish to me now

File: src/bank.rs [hunk 1]
```diff
@@ -206,45 +317,70 @@ impl Bank {
...
+        let rolled_back_pubkeys: Vec<Pubkey> = self
+            .accounts
+            .read()
+            .unwrap()
+            .accounts
+            .keys()
+            .map(|pubkey| pubkey.clone())
+            .collect();
+
+        self.accounts.write().unwrap().rollback();
```
Comment (carllin) [inline src/bank.rs]: Will there be  issues  if there are ongoing transactions being processed in execute_and_commit_transactions(), which loads the account, operates on it, and then stores it at some arbitrary point in the future?

**Comment by carllin** [comment]
This is awesome!

File: src/bank.rs [hunk 2]
```diff
@@ -1228,18 +1351,22 @@ impl Bank {
...
     }
 
     /// Hash the `accounts` HashMap. This represents a validator's interpretation
-    ///  of the ledger up to the `last_id`, to be sent back to the leader when voting.
+    ///  of the delta of the ledger since the last vote and up to now
     pub fn hash_internal_state(&self) -> Hash {
         let mut ordered_accounts = BTreeMap::new();
-        for (pubkey, account) in self.accounts.read().unwrap().iter() {
+
+        // only hash internal state of the part being voted upon, i.e. since last
```
Comment (rob-solana) [inline src/bank.rs]: @garious @aeyakovenko is this bit of code is correct?

It implies that we expect all nodes to implement this voting hash functionality the same way, i.e. that all nodes voting on this segment of the ledger have **exactly the same checkpoint delta**.  I think it's right, but would appreciate a sanity check.

Comment (garious) [reply src/bank.rs]: Correct that the algorithm to generate the voting hash is part of the network protocol. Please make sure this particular flavor makes it into an RFC sooner than later. Once there, I'll add it to the architecture doc.

**Pull request merged**